// @upstash/redis/script/platforms/cloudflare.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Redis = void 0;
const core = __importStar(require("../pkg/redis.js"));
const error_js_1 = require("../pkg/error.js");
/**
 * Serverless redis client for upstash.
 */
class Redis extends core.Redis {
    /**
     * Create a new redis client
     *
     * @example
     * ```typescript
     * const redis = new Redis({
     *  url: "<UPSTASH_REDIS_REST_URL>",
     *  token: "<UPSTASH_REDIS_REST_TOKEN>",
     * });
     * ```
     */
    constructor(config) {
        const client = defaultRequester({
            baseUrl: config.url,
            headers: { authorization: `Bearer ${config.token}` },
        });
        super(client, {
            automaticDeserialization: config.automaticDeserialization,
        });
    }
    /*
     * Create a new Upstash Redis instance from environment variables on cloudflare.
     *
     * This tries to load `UPSTASH_REDIS_REST_URL` and `UPSTASH_REDIS_REST_TOKEN` from
     * the global namespace
     *
     * If you are using a module worker, please pass in the `env` object.
     * ```ts
     * const redis = Redis.fromEnv(env)
     * ```
     *
     */
    static fromEnv(env) {
        // @ts-ignore These will be defined by cloudflare
        const url = env?.UPSTASH_REDIS_REST_URL ?? UPSTASH_REDIS_REST_URL;
        // @ts-ignore These will be defined by cloudflare
        const token = env?.UPSTASH_REDIS_REST_TOKEN ?? UPSTASH_REDIS_REST_TOKEN;
        if (!url) {
            throw new Error("Unable to find environment variable: `UPSTASH_REDIS_REST_URL`. Please add it via `wrangler secret put UPSTASH_REDIS_REST_URL`");
        }
        if (!token) {
            throw new Error("Unable to find environment variable: `UPSTASH_REDIS_REST_TOKEN`. Please add it via `wrangler secret put UPSTASH_REDIS_REST_TOKEN`");
        }
        return new Redis({ url, token });
    }
}
exports.Redis = Redis;
function defaultRequester(config) {
    return {
        request: async function (req) {
            if (!req.path) {
                req.path = [];
            }
            const res = await fetch([config.baseUrl, ...req.path].join("/"), {
                method: "POST",
                headers: { "Content-Type": "application/json", ...config.headers },
                body: JSON.stringify(req.body),
                keepalive: true,
            });
            const body = (await res.json());
            if (!res.ok) {
                throw new error_js_1.UpstashError(body.error);
            }
            return body;
        },
    };
}
