// @upstash/redis/esm/platforms/fastly.js
import * as core from "../pkg/redis.js";
import { UpstashError } from "../pkg/error.js";
/**
 * Serverless redis client for upstash.
 */
export class Redis extends core.Redis {
    /**
     * Create a new redis client
     *
     * @example
     * ```typescript
     *    const redis = new Redis({
     *        url: "<UPSTASH_REDIS_REST_URL>",
     *        token: "<UPSTASH_REDIS_REST_TOKEN>",
     *        backend: "upstash-db",
     *    });
     * ```
     */
    constructor(config) {
        const client = defaultRequester({
            baseUrl: config.url,
            headers: { authorization: `Bearer ${config.token}` },
            backend: config.backend,
        });
        super(client, {
            automaticDeserialization: config.automaticDeserialization,
        });
    }
}
function defaultRequester(config) {
    return {
        request: async function (req) {
            if (!req.path) {
                req.path = [];
            }
            const res = await fetch([config.baseUrl, ...req.path].join("/"), {
                method: "POST",
                headers: { "Content-Type": "application/json", ...config.headers },
                body: JSON.stringify(req.body),
                keepalive: true,
                // @ts-expect-error fastly requires `backend`
                backend: config.backend,
            });
            const body = (await res.json());
            if (!res.ok) {
                throw new UpstashError(body.error);
            }
            return body;
        },
    };
}
