// @upstash/redis/esm/platforms/nodejs.js
// deno-lint-ignore-file
import * as core from "../pkg/redis.js";
import { UpstashError } from "../pkg/error.js";
// @ts-ignore Deno can't compile
// import https from "https";
// @ts-ignore Deno can't compile
// import http from "http";
import "isomorphic-fetch";
/**
 * Serverless redis client for upstash.
 */
export class Redis extends core.Redis {
    constructor(configOrRequester) {
        if ("request" in configOrRequester) {
            super(configOrRequester);
            return;
        }
        const client = defaultRequester({
            baseUrl: configOrRequester.url,
            headers: { authorization: `Bearer ${configOrRequester.token}` },
            // agent: configOrRequester.agent,
        });
        super(client, {
            automaticDeserialization: configOrRequester.automaticDeserialization,
        });
    }
    /**
     * Create a new Upstash Redis instance from environment variables.
     *
     * Use this to automatically load connection secrets from your environment
     * variables. For instance when using the Vercel integration.
     *
     * This tries to load `UPSTASH_REDIS_REST_URL` and `UPSTASH_REDIS_REST_TOKEN` from
     * your environment using `process.env`.
     */
    static fromEnv(config) {
        // @ts-ignore process will be defined in node
        if (typeof process?.env === "undefined") {
            throw new Error('Unable to get environment variables, `process.env` is undefined. If you are deploying to cloudflare, please import from "@upstash/redis/cloudflare" instead');
        }
        // @ts-ignore process will be defined in node
        const url = process?.env["UPSTASH_REDIS_REST_URL"];
        if (!url) {
            throw new Error("Unable to find environment variable: `UPSTASH_REDIS_REST_URL`");
        }
        // @ts-ignore process will be defined in node
        const token = process?.env["UPSTASH_REDIS_REST_TOKEN"];
        if (!token) {
            throw new Error("Unable to find environment variable: `UPSTASH_REDIS_REST_TOKEN`");
        }
        return new Redis({ url, token, ...config });
    }
}
function defaultRequester(config) {
    return {
        request: async function (req) {
            if (!req.path) {
                req.path = [];
            }
            const res = await fetch([config.baseUrl, ...req.path].join("/"), {
                method: "POST",
                headers: { "Content-Type": "application/json", ...config.headers },
                body: JSON.stringify(req.body),
                keepalive: true,
                // @ts-ignore
                agent: config.agent,
            });
            const body = (await res.json());
            if (!res.ok) {
                throw new UpstashError(body.error);
            }
            return body;
        },
    };
}
