// @upstash/redis/script/pkg/http.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpClient = void 0;
const error_js_1 = require("./error.js");
class HttpClient {
    constructor(config) {
        Object.defineProperty(this, "baseUrl", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "headers", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "options", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.baseUrl = config.baseUrl.replace(/\/$/, "");
        this.headers = { "Content-Type": "application/json", ...config.headers };
        this.options = config.options;
    }
    async request(req) {
        if (!req.path) {
            req.path = [];
        }
        const requestOptions = {
            method: "POST",
            headers: this.headers,
            body: JSON.stringify(req.body),
            keepalive: true,
            /**
             * Fastly specific
             */
            backend: this.options?.backend,
        };
        // fetch is defined by isomorphic fetch
        // eslint-disable-next-line no-undef
        const res = await fetch([this.baseUrl, ...req.path].join("/"), requestOptions);
        const body = (await res.json());
        if (!res.ok) {
            throw new error_js_1.UpstashError(body.error);
        }
        return body;
    }
}
exports.HttpClient = HttpClient;
